transaction/core/tests/input_rules.rs

```diff
@@ -63,25 +72,290 @@ fn test_input_rules_verify_max_tombstone() {
+            let tx_out_shared_secret =
+                get_tx_out_shared_secret(alice.view_private_key(), &decompressed_tx_pub);
+            let amount_shared_secret =
+                MaskedAmount::compute_amount_shared_secret(block_version, &tx_out_shared_secret)
+                    .unwrap();
+            let (amount, _) = txo
+                .masked_amount
+                .as_ref()
+                .unwrap()
+                .get_value_from_amount_shared_secret(&amount_shared_secret)
+                .unwrap();
+            assert_eq!(amount.value, 1000);
+            assert_eq!(*amount.token_id, 0);
+            RevealedTxOut {
+                tx_out: txo.clone(),
+                amount_shared_secret: amount_shared_secret.to_vec(),
+            }
+        })
+        .collect();
+
+    // Add a fractional output, by doubling one of the revealed tx outs.
+    // So this means we are filling the 1/2 of the request.
+    // Before we add a fractional change output, this is ill-formed, we can only
+    // verify fractional output rules if there is a fractional change output,
+    // because that is how we deduce the fill fraction.
+    get_first_rules_mut(&mut tx)
+        .fractional_outputs
+        .push(change_committed_amount(
+            &revealed_tx_outs[1],
+            Amount::new(2000, 0.into()),
+        ));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::FractionalOutputsNotExpected)
+    );
+
+    // Add a fractional input, also by doubling one of the revealed tx outs.
+    // This means the fill fraction is 1/2, which we are satisfying.
+    get_first_rules_mut(&mut tx).fractional_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(2000, 0.into()),
+    ));
+
+    // Check that the Tx is following input rules
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Lets try imposing a limit on the change value
+    // This is larger than 1000 so we should still be valid.
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 1010;
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Lets try imposing a smaller limit that should cause things to fail
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 990;
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::RealChangeOutputAmountExceedsLimit)
+    );
+
+    // Remove limit altogether, we should be valid again
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 0;
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Change the frational change output to be 3000. This means the implied fill
+    // fraction is now 2/3, since the real change output only returns
+    // 1/3 of this. The tx should then be invalid because we are only filling 1/2 of
+    // the fractional output that was required and not 2/3 of it.
```

Comment:
:x: I think this is worded incorrectly and the logic may possibly be backwards
Alice has put up 3000 of token X.  Bob is taking 1000 of token X (txout_0)
Alice requested 2000 of token Y, Bob is providing 1000 of token Y (txout_1)

While this is disadvantageous for Bob, in the end he's the one making this overall TX

There should also be a test for the boundary conditions, there is quite a bit of room between 1/3 and 1/2

Reply:
>  I think this is worded incorrectly and the logic may possibly be backwards

Can you clarify exactly what you think is wrong and what you think is backwards?

I'd prefer to use the precise terms "fractional output" and "fractional change output" rather than "put up" and "requested" because these terms are ambiguous. It's important to note that the amount Alice has "put up" is the value of the input underlying the SCI, which should, but is not technically required, to match the value of the fractional change output. And in this test, these things do not match, but it isn't important for testing the functionality we are interested in here.

> There should also be a test for the boundary conditions, there is quite a bit of room between 1/3 and 1/2

I agree, but I think it's also valuable to change one thing, test, change another thing, test, to simply get more test coverage than we would have otherwise. I can make sure we test the boundary though as well.

Reply:
> but I think it's also valuable to change one thing, test, change another thing, test
:100: